File: benchmark-csharp/src/main/csharp/EnumTest.cs
using System;
using System.Collections.Generic;
using System.Text;

sealed class EnumTest
{
    
    /**
     * NB_TESTS
     */
    private static int NB_TESTS = 5;

    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    private static int NB_OF_EXCLUSION_MIN_MAX = 2;

    /**
     * NB_CREATION_TESTS
     */
    private static int NB_ENUM_TESTS = 2 * 1000 * 1000;

    /**
     * NB_CREATION_TESTS
     */
    private static int NB_ENUM_PARSE_SWITCH_TESTS = 10 * 1000 * 1000;

    public enum Number
    {
        ONE, TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE, TEN
    }

    /**
     * averageTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    public static long averageTimeWithoutMinMax(List<long> executionTimes, int numberOfMinMaxToRemove)
    {
        long average = 0;
		int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0;
		if (minMax != 0) 
		{
        	executionTimes.Sort();
		}
        for (int i = minMax; i < executionTimes.Count - minMax; i++)
        {
            average += executionTimes[i];
        }
        average = average / (executionTimes.Count - minMax * 2);
        return average;
    }


    /**
     * relativeDeviationTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    public static double relativeDeviationTimeWithoutMinMax(List<long> executionTimes, int numberOfMinMaxToRemove)
    {
        long averageTimeWithoutMinMaxVariable = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        long deviation = 0;
		int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0;
		if (minMax != 0) 
		{
        	executionTimes.Sort();
		}
        for (int i = minMax; i < executionTimes.Count - minMax; i++)
        {
            deviation += (long)Math.Pow(executionTimes[i] - averageTimeWithoutMinMaxVariable, 2);
        }
        return ((int)((Math.Sqrt(deviation / (executionTimes.Count - 2 * minMax)) / averageTimeWithoutMinMaxVariable * 100) * 100)) / 100d;
    }
    
    

    
    /**
     * testCurrentTime.
     *
     * @return
     */
    public static long testEnumGetValues()
    {
	    DateTime start = DateTime.Now;
        for (int i = NB_ENUM_TESTS; i != 0; i--)
	    {
            Number[] numbers = (Number[]) Enum.GetValues(typeof(Number));
	    }
	    DateTime end = DateTime.Now;
	    TimeSpan executionTime = end - start;
        Console.WriteLine("[EnumTest], Invoke of " + NB_ENUM_TESTS + " Number[] numbers = (Number[]) Enum.GetValues(typeof(Number)),, snapshot time,"
		    + executionTime.TotalMilliseconds);
	    return (long) executionTime.TotalMilliseconds;
    }

    /**
 * testCurrentTime.
 *
 * @return
 */
    public static long testEnumParseAndSwitch()
    {
        int count = 0;
        DateTime start = DateTime.Now;
        for (int i = NB_ENUM_PARSE_SWITCH_TESTS; i != 0; i--)
        {
            Number number = (Number) Enum.Parse(typeof(Number), "ONE");
            switch (number)
            {
                case Number.ONE:
                    count = count + 1;
                    break;
                case Number.TWO:
                    count = count + 2;
                    break;
                case Number.THREE:
                    count = count + 3;
                    break;
                case Number.FOUR:
                    count = count + 4;
                    break;
                case Number.FIVE:
                    count = count + 5;
                    break;
                case Number.SIX:
                    count = count + 6;
                    break;
                case Number.SEVEN:
                    count = count + 7;
                    break;
                case Number.EIGHT:
                    count = count + 8;
                    break;
                case Number.NINE:
                    count = count + 9;
                    break;
                case Number.TEN:
                    count = count + 10;
                    break;
                default:
                    break;
            }
        }
        DateTime end = DateTime.Now;
        TimeSpan executionTime = end - start;
        Console.WriteLine("[EnumTest], Invoke of " + NB_ENUM_PARSE_SWITCH_TESTS + "  Number number = (Number) Enum.Parse(typeof(Number)  \"ONE\") switch (number)..., count=" + count + ", snapshot time,"
            + executionTime.TotalMilliseconds);
        return (long)executionTime.TotalMilliseconds;
    }
    
	
	    /**
     * main.
     *
     * @param args
     */
    public static void main(string[] args)
    {
	    int nbTests = (args != null && args.Length >= 1) ? int.Parse(args[0]) : NB_TESTS;
	    int nbOfExclusionMinMax = (args != null && args.Length >= 2) ? int.Parse(args[1]) : NB_OF_EXCLUSION_MIN_MAX;

	    List<long> executionTimes = new List<long>(nbTests);
        
	    for (int i = nbTests; i != 0; i--)
            executionTimes.Add(EnumTest.testEnumGetValues());
        executionTimes.Sort();
        Console.WriteLine("[EnumTest], Invoke of " + NB_ENUM_TESTS + " Number[] numbers = (Number[]) Enum.GetValues(typeof(Number)),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes[0] + ", max time," + executionTimes[executionTimes.Count - 1]
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
	    executionTimes.Clear();
        

        for (int i = nbTests; i != 0; i--)
            executionTimes.Add(EnumTest.testEnumParseAndSwitch());
        executionTimes.Sort();
        Console.WriteLine("[EnumTest], Invoke of " + NB_ENUM_PARSE_SWITCH_TESTS + " Number number = (Number) Enum.Parse(typeof(Number)  \"ONE\") switch (number)...,, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes[0] + ", max time," + executionTimes[executionTimes.Count - 1]
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.Clear();
    }

}

File: benchmark-java/src/main/java/EnumTest.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * EnumTest.
 * 
 * @author Alexandre
 * 
 */
public final class EnumTest
{
    /**
     * NB_TESTS
     */
    private static final int NB_TESTS                   = 5;

    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    private static final int NB_OF_EXCLUSION_MIN_MAX    = 2;

    /**
     * NB_ENUM_TESTS
     */
    private static final int NB_ENUM_TESTS              = 2 * 1000 * 1000;

    /**
     * NB_ENUM_PARSE_SWITCH_TESTS
     */
    private static final int NB_ENUM_PARSE_SWITCH_TESTS = 10 * 1000 * 1000;


    
    /**
     * Added to achieve to compile with the groovy compiler
     */

    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum ONE = EnumTest.NumberEnum.ONE;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum TWO = EnumTest.NumberEnum.TWO;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum THREE = EnumTest.NumberEnum.THREE;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum FOUR = EnumTest.NumberEnum.FOUR;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum FIVE = EnumTest.NumberEnum.FIVE;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum SIX = EnumTest.NumberEnum.SIX;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum SEVEN = EnumTest.NumberEnum.SEVEN;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum EIGHT = EnumTest.NumberEnum.EIGHT;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum NINE = EnumTest.NumberEnum.NINE;
    @SuppressWarnings("unused")
    private static final EnumTest.NumberEnum TEN = EnumTest.NumberEnum.TEN;


    /**
     * averageTimeWithoutMinMax.
     * 
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static long averageTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        long average = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            average += executionTimes.get(i);
        }
        average = average / (executionTimes.size() - minMax * 2);
        return average;
    }
    
    
    /**
     * relativeDeviationTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static double relativeDeviationTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        final long averageTimeWithoutMinMax = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        long deviation = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            deviation += Math.pow(executionTimes.get(i) - averageTimeWithoutMinMax, 2);
        }
        return ((int) ((Math.sqrt(deviation / (executionTimes.size() - 2 * minMax)) / averageTimeWithoutMinMax * 100) * 100)) / 100d;
    }    

    /**
     * testCurrentTime.
     * 
     * @return
     */
    public static long testEnumGetValues()
    {
        long start = System.currentTimeMillis();
        int count = 0;
        start = System.currentTimeMillis();
        for (int i = NB_ENUM_TESTS; i != 0; i--)
        {
            // EnumTest.NumberEnum instead of NumberEnum because of groovy
            final EnumTest.NumberEnum[] numbers = EnumTest.NumberEnum.values();
            count += numbers.length;
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[EnumTest], Invoke of " + NB_ENUM_TESTS + " final Number[] numbers = Number.values(), count=" + count
                + ", snapshot time," + executionTime);
        return executionTime;
    }

    /**
     * testEnumParseAndSwitch.
     * 
     * @return
     */
    public static long testEnumParseAndSwitch()
    {
        long start = System.currentTimeMillis();
        int count = 0;
        start = System.currentTimeMillis();
        for (int i = NB_ENUM_PARSE_SWITCH_TESTS; i != 0; i--)
        {
            // EnumTest.NumberEnum instead of NumberEnum because of groovy
            final EnumTest.NumberEnum number = EnumTest.NumberEnum.valueOf("ONE");
            
            switch (number)
            {
            case ONE:
                count = count + 1;
                break;
            case TWO:
                count = count + 2;
                break;
            case THREE:
                count = count + 3;
                break;
            case FOUR:
                count = count + 4;
                break;
            case FIVE:
                count = count + 5;
                break;
            case SIX:
                count = count + 6;
                break;
            case SEVEN:
                count = count + 7;
                break;
            case EIGHT:
                count = count + 8;
                break;
            case NINE:
                count = count + 9;
                break;
            case TEN:
                count = count + 10;
                break;
            default:
                break;
            }
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[EnumTest], Invoke of " + NB_ENUM_PARSE_SWITCH_TESTS + " final Number number = Number.valueOf(\"ONE\") switch (number)..., count="
                + count + ", snapshot time," + executionTime);
        return executionTime;
    }

    
    /**
     * main.
     * 
     * @param args
     */
    @SuppressWarnings("boxing")
    public static void main(final String[] args)
    {
        final int nbTests = (args != null && args.length >= 1 && args[0] != null) ?  Integer.valueOf(args[0]): NB_TESTS;
        final int nbOfExclusionMinMax = (args != null && args.length >= 2 && args[1] != null) ? Integer.valueOf(args[1]) : NB_OF_EXCLUSION_MIN_MAX;

        final List<Long> executionTimes = new ArrayList<Long>(nbTests);

        for (int i = nbTests; i != 0; i--)
            executionTimes.add(EnumTest.testEnumGetValues());
        System.out.println("[EnumTest], Invoke of " + NB_ENUM_TESTS + " final Number[] numbers = Number.values(),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();

        for (int i = nbTests; i != 0; i--)
            executionTimes.add(EnumTest.testEnumParseAndSwitch());
        System.out.println("[EnumTest], Invoke of " + NB_ENUM_PARSE_SWITCH_TESTS + " final Number[] numbers = Number.values(),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();

    }

    /**
     * NUMBER.
     * 
     * @author anavarro122404 - 25 Oct 2007
     *
     *
     *
     */
    public static enum NumberEnum
    {
        ONE, TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE, TEN
    }
}

